healthz/healthz.proto

```diff
@@ -0,0 +1,65 @@
+//
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Generic Network Operation Interface, GNOI, defines a set of RPC's used for
+// the operational aspects of network targets. These services are meant to be
+// used in conjunction with GNMI for all target state and operational aspects
+// of a network target. The gnoi.system.Service is the only mandatory vendor
+// implementation.
+
+syntax = "proto3";
+
+
+package gnoi.healthz;
+
+import "github.com/openconfig/gnoi/types/types.proto";
+import "google/protobuf/any.proto";
+
+option go_package = "github.com/openconfig/gnoi/healthz";
+
+option (types.gnoi_version) = "1.0.0";
+
+// The Healthz service provides access to the status of a path on the
+// system. Addtitionally it allows the implementor to provide path specific
+// diagnositic data into the status return.
+//
+// Healthz is expected to work in conjunction with the component OC model.
+service Healthz {
+  // Get will get health status for a gNMI path.  If no status is available for
+  // the requested path an error will be returned.
+  rpc Get(GetRequest) returns (GetResponse) {}
```

Comment:
Glad to see this. This service seems like something we've really wanted better defined for the optical space for a while. One of our main use cases is collecting a state-dump from the device. Essentially, most optical vendors offer an ability in which the device internally compiles a large file (i.e. 100MB) either for the entire device or a specific component and then we send that to the vendor.

Today, we are using gNOI.File.TransferToRemote in a bit of a hacky way (using a magic path essentially from the vendor) to initiate the compiling of the state-dump and transferring it. We are planning to switch this to use gNOI.File.Get in upcoming releases so we can stream the data instead.

With that background, would it make sense to have this RPC be more specific so that we could expand this service with additional use cases? One thought would be to switch Get --> Component, then add a new RPC which could stream a file back (perhaps in a future PR).

service Healthz {
  rpc Component(ComponentRequest) returns (ComponentResponse) {}
  rpc DebugFile(DebugFileRequest) returns (stream DebugFileResponse) {}
}

If this use case doesn't fit well, no problem, just wanted to see. Thanks.

Reply:
the proto.Any definition allows for the vendor to decide what types of information are relevant - this could be "debug files" or any other data.

as to if we wanted to allow for "chunking" of the data seems like chunking it by component vs. providing a streaming solution would be easier but we could iterate on that approach if you really feel building a single "response" is too big for a general use case

Reply:
In the case of just providing a large file back, I think wrapping it in a proto.Any might not be ideal here as we or a vendor would have to define a new proto definition which really only needs to contain one field that would be of type bytes. Certainly do-able, but it's potentially an avoidable extra layer. I think a streaming solution that streams bytes directly (and concludes with a hash for verification) could be a nice simple solution. In my case, these are just large files that we have no insight into what the details are (we are just getting a file to pass along to vendor developers).

I don't mean to hold up this PR, so feel free to proceed as is, just wanted to share as this has been a use case I've been wanting to see get addressed and this service seems like it might be the right place :)

Reply:
we don't need to define the point would the vendor can decide what data is relevant and then package it however they want.

Streaming doesn't really "fix anything" as you still have to have the data locally and then send it over the rpc.. the only "optimization" is in the memory needed to serialize the data in memory then write it - we already have a file operation for streaming a file off a box.  The proto they define could very well be to be file pointer that can be used to then fetch the file over the file api.

The other implications would be a component by component health check, would be you can say only get me the diagnostic information relevant to the component vs. "show tech" which dumps state about everything.